libkbfs/key_manager.go

```diff
@@ -124,19 +125,80 @@ func (km *KeyManagerStandard) getTLFCryptKey(ctx context.Context,
 		return TLFCryptKey{}, err
 	}
 
-	localMakeRekeyReadError := func() error {
-		return makeRekeyReadError(ctx, km.config, kmd, keyGen, uid, username)
+	crypto := km.config.Crypto()
+
+	origKeyGen := keyGen
+	notPerDeviceEncrypted :=
+		!kmd.AreHistoricTLFCryptKeysPerDeviceEncrypted() && keyGen < kmd.LatestKeyGeneration()
+	if notPerDeviceEncrypted {
+		// If the crypt key at this generation isn't per-device encrypted
+		// we need to get the latest crypt key to decrypt it symmetrically.
+		keyGen = kmd.LatestKeyGeneration()
+		// look in the cache first
+		kcache := km.config.KeyCache()
+		tlfCryptKey, err = kcache.GetTLFCryptKey(tlfID, keyGen)
+		switch err := err.(type) {
+		case nil:
+			break
+		case KeyCacheMissError:
+			break
+		default:
+			return TLFCryptKey{}, err
+		}
+	}
+
+	if (tlfCryptKey == TLFCryptKey{}) {
+		clientHalf, serverHalfID, cryptPublicKey, err :=
```

Comment:
So I think this code could be refactored differently to be clearer; one problem is that this function call means different things depending on v2 or v3. For v2, this retrieves the params for the `TLFCryptKey` for `keyGen`, but for v3, this actually retrieves the params for the latest `TLFCryptKey` (the `keyGen` parameter is ignored), which is then used below, which is kinda confusing.

I think instead a better way to structure this would be:
1) make `getTLFCryptKeyParams` for V3 return a specific error if `keyGen < latestKeyGen`.
2) Have the flow of this function be something like:

a) Try getting the `TLFCryptKey` for `keyGen` from the cache. If that works, return it.
b) Otherwise, call `getTLFCryptKeyParams` with `keyGen`. If that works, get the server half and return the unmasked key. (v2 flow)
c) Otherwise, if `getTLFCryptKeyParams` returns the error from 1) above, then call it with `latestKeyGen`, get the server half, get the unmasked latest key, and use that to call `GetTLFCryptKey`.

As a bonus, this won't need the awkwardly-named `AreHistoricTLFCryptKeysPerDeviceEncrypted` function.

What do you think?